kubetest/anywhere.go

```diff
@@ -116,8 +122,23 @@ func newKubernetesAnywhere(project, zone string) (*kubernetesAnywhere, error) {
 		Zone:              zone,
 	}
 
-	if err := k.writeConfig(); err != nil {
-		return nil, err
+	k.multiClusters = multiClusters
+	if !k.multiClusters.Enabled() {
+		if err := k.writeConfig(); err != nil {
+			return nil, err
+		}
+	} else {
+		for _, cluster := range k.multiClusters.clusters {
```

Comment:
The way this is implemented, all of the logic of how to handle multiCluster scenarios is now injected directly into the kubernetes-anywhere deployment method, and muddying all of its methods with `if multiClusters.Enabled()` checks to either treat the cluster as a single cluster or multiple. Not only does this make every method a big `if-then-else` block of single vs. multiCluster, but that also means that if we ever want to add multiCluster support for kops, gke, etc., we have to go through this entire process again to make all of their methods multiCluster-aware.

I would love to see this behavior made more composeable with the existing deployment methods, even if the other deployments aren't fully support it yet. What do you think about this approach:

1) Make a new `MultiClusterDeployer` that is a new implementation of the `deployer` interface.
2) Its constructor accepts an array of other `deployer`s.
3) All of its `Up()`, `Down()`, etc. methods are just for-loops to delegate the call to each of the N `deployer`s it was constructed with, probably in parallel goroutines, and probably aggregating the results for things like `IsUp()`.
4) In `main.go`, when multiCluster is enabled and the deployment is kubernetes-anywhere, you just construct N different kubernetesAnywhere deployers and wrap them in your new `MultiClusterDeployer`.

You'll still need to make minor modifications in `anywhere.go` to make the config file location a field instead of hardcoded, but it can mostly go untouched. Then, if we ever want to add multiCluster support for a different `deployer`, all we need is a tiny bit of logic in `main.go` that knows how to construct N instances of the `deployer` (each with the correct arguments to differentiate them) and then wrap them with a `MultiClusterDeployer` and we're done. There may be other minimal changes to the other `deployer` to make sure that multiple instances don't clash with other, but otherwise it can stay completely unaware of the multiCluster needs.

What do you think of that approach?

Reply:
I agree with your proposal and thats the clean and extensible way to do if we want to deploy clusters serially. I was more biased towards deploying clusters in parallel. To overcome the amount of time it takes to deploy 3 clusters in federation, we had to do a lot of circus, which also brought additional issues with it. So from federation perspective it is very important that we do deployments in parallel. Although doing it serially means maintaining the status quo. (We wanted to solve this problem along with changing the deployment method).

With your proposed solution, deploying in parallel is not possible in the current state. for e.g: `Up()` method in deployer mostly calls the `finishRunning()` function which is not possible to call in parallel by multiple goroutines for 2 reasons:
- the logging is happening directly to stdout, and its not good that multiple goroutines write to stdout parallely.
- the timeout and termination events are handled in finishRunning() function. if we have multiple goroutines run the finishRunning() function, only one of them will get the event and handle it but rest of them will continue to run, which is not desirable.
I am little hesitant to change the behavior of `finishRunning()` function as it is used in so many places in kubetest.

Alternatively i could just create a separate deployer for multi-clusters in k8s-anywhere, which is not as elegant as you proposed, but the code will look cleaner.

Reply:
I completely understand the desire for parallelism, and I don't think the two approaches are mutually exclusive. You've already written `finishRunningParallel()`, so you've already figured out how to buffer the stdout/stderr of each child process so that you can separate their output instead of having lines interwoven in real-time, and how to handle interruption signals so that you can clean up all of the parallel processes.

Here's how I think we can have the best of both worlds: since all of the existing `deployer`s are already using `finishRunning()`, and you mentioned that you don't want to change the existing implementation for fear of breaking them, I would recommend a solution where you're able to change the implementation dynamically at runtime.

The quick-and-dirty way that I think would be sufficient is to just make `finishRunning` a variable that can be overridden:

    var finishRunning = func(cmd *exec.Cmd) error {

With this change, none of the existing calls to `finishRunning()` need to change at all, but since `finishRunning` is now a pointer, you can dynamically update it to point to a different, parallel-friendly implementation when you're running with multiClusters enabled. All of the existing jobs that are using the other deployers without multiClusters will continue to work exactly as they do today, but you can thoroughly test your parallelFinishRunning implementation with kubernetes-anywhere + multiClusters, which is the only time it would be used.

We currently use this method for [getKube()](https://github.com/kubernetes/test-infra/blob/eec243bae2e7ad2d6e2cd37b61d8bdc8cdd1eb11/kubetest/extract.go#L211) so that we can use a substituted implementation during [unit tests](https://github.com/kubernetes/test-infra/blob/b478098499a8bc30b67b9b8fa334b46968ab4505/kubetest/extract_test.go#L118). I believe this is adequate because you'll never want different `deployer`s using a mix of different implementations during the lifetime of the process: you'll want everyone using one implementation or the other.

A cleaner strategy that would require more refactoring would be to hide `finishRunning` behind an interface, and pass the implementation to use to each `deployer`'s constructor. In `main.go`, each `deployer` would get constructed with the standard implementation of `finishRunning` unless multiCluster is on, and then you would just swap the implementation out for the `parallelFinishRunning` variant:

    finishRunning = parallelFinishRunning

The only other catch to these two approaches is that not all parallel commands are known at the time you start executing them, so you would need to track them dynamically as more instances of `parallelFinishRunning` get invoked. This seems achievable with a small tweak to your existing implementation, moving the `wg` WaitGroup outside of the function into a package-level or object-level field and calling `wg.Add(1)` every time `parallelFinishRunning` is called, and similarly tracking all of the existing in-flight parallel commands being run.

Is that a palatable solution?

Reply:
I think an interface and a waitgroup should be used for this :+1:

Reply:
@pipejakob, Thanks for the detailed layout and i think it is feasible. i would give it a try in couple of days.